src/org/ods/orchestration/util/DocumentHistory.groovy

```diff
@@ -227,6 +212,21 @@ class DocumentHistory {
         sortDocHistories(dhs).reverse()
     }
 
+    // Find the previous environment where the document was generated, if it exists.
+    // Otherwise, use the target environment.
+    @NonCPS
+    private getSourceEnvironment(String targetEnvironment, String documentName) {
+        def documentType = LeVADocumentUtil.getTypeFromName(documentName) as String
+        def environment = null
+        def envs = ['D', 'Q', 'P']
```

Comment:
@jafarre-viewnext should `envs` be defined in-line? Don't we have a global reference already?

Reply:
I extracted the environments into an Enum